internal/testutil/redis.go

```diff
@@ -85,6 +85,96 @@ func WithTestRedis(useTLS bool, handler func(rawURL string) error) error {
 	return e
 }
 
+// WithTestRedisCluster creates a new redis cluster 3 node cluster.
+func WithTestRedisCluster(handler func(rawURL string) error) error {
+	ctx, clearTimeout := context.WithTimeout(context.Background(), maxWait)
+	defer clearTimeout()
+
+	// uses a sensible default on windows (tcp/http) and linux/osx (socket)
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		return err
+	}
+
+	redises := make([]*dockertest.Resource, 3)
+	for i := range redises {
+		conf := "cluster-enabled yes\ncluster-config-file nodes.conf"
+		r, err := pool.RunWithOptions(&dockertest.RunOptions{
+			Hostname:   fmt.Sprintf("redis%d", i),
+			Repository: "redis",
+			Tag:        "6",
+			Entrypoint: []string{
+				"/bin/bash", "-c",
+				`echo "` + conf + `" >/tmp/redis.conf && chmod 0777 /tmp/redis.conf && exec docker-entrypoint.sh /tmp/redis.conf`,
+			},
+			ExposedPorts: []string{
+				"6379/tcp",
+				"26379/tcp",
+			},
+		})
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+		_ = r.Expire(uint(maxWait.Seconds()))
+
+		go func() {
+			_ = pool.Client.Logs(docker.LogsOptions{
+				Context:      ctx,
+				Stderr:       true,
+				Stdout:       true,
+				Follow:       true,
+				Timestamps:   true,
+				Container:    r.Container.ID,
+				OutputStream: os.Stderr,
+				ErrorStream:  os.Stderr,
+			})
+		}()
+
+		redises[i] = r
+	}
+	addrs := make([]string, 3)
+	for i, r := range redises {
+		addrs[i] = net.JoinHostPort(
+			r.Container.NetworkSettings.IPAddress,
+			"6379",
+		)
+	}
+
+	for _, addr := range addrs {
+		err := pool.Retry(func() error {
+			options, err := redis.ParseURL(fmt.Sprintf("redis://%s/0", addr))
+			if err != nil {
+				return err
+			}
+
+			client := redis.NewClient(options)
+			defer client.Close()
+
+			return client.Ping(ctx).Err()
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// join the nodes to the cluster
+	err = bootstrapRedisCluster(ctx, redises)
+	if err != nil {
+		return err
+	}
+
+	e := handler(fmt.Sprintf("redis+cluster://%s", strings.Join(addrs, ",")))
+
+	for _, r := range redises {
+		if err := pool.Purge(r); err != nil {
+			return err
+		}
+	}
```

Comment:
maybe wrap this into `defer func()` ?

Reply:
idk how important is it to purge docker test pool; but also maybe try to clean up as many containers as possible and collect errors into `multierror.Error`

Reply:
we already defer a purge when the resource is created